New package: py-fastfold

meyersbs opened the pull request:
This was not fun to package: [GitHub](https://github.com/hpcaitech/FastFold/tree/0.2.0). It doesn't exist in PyPI and the preferred installation method is via conda environment. As a result, the `setup.py` is incomplete.

Notes:
- I manually tested imports for dependencies listed in the conda `environment.yml` until I got everything working.
- The conda environment lists `PyYAML`, `tqdm`, `typing-extensions`, and `pyarrow`, but those aren't used anywhere in the source code, so I left them out of the spack package.
- I also had to pull a few dependencies from the README, and add some dependencies that aren't even listed in the conda environment because of import errors.

Import test:
```
Python 3.8.7 (default, Dec 2 2022, 14:27:46) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastfold; from fastfold.config import * ; from fastfold.workflow import * ; from fastfold.utils import * ; from fastfold.relax import * ; from fastfold.model import * ; from fastfold.distributed import * ; from fastfold.data import * ; from fastfold.common import * ; from fastfold.data.tools import * ; from fastfold.model.fastnn import * ; from fastfold.model.hub import * ; from fastfold.model.nn import * ; from fastfold.model.nn.embedders import * ; from fastfold.model.nn.msa import * ; from fastfold.model.nn.dropout import * ; from fastfold.model.hub.alphafold import * ; from fastfold.model.fastnn.embedders import * ; from fastfold.model.fastnn.evoformer import * ; from fastfold.utils.geometry import * ; from fastfold.utils.geometry.rigid_matrix_vector import * ; from fastfold.workflow.factory import * ; from fastfold.workflow.template import *; from fastfold.workflow.factory.hhsearch import *; from fastfold.data.data_modules import *; from fastfold.relax.amber_minimize import *
[12/02/22 15:42:52] WARNING  colossalai - root - WARNING: Triton is not available, fallback to old kernel.                                                                                                                                                    
                    WARNING  colossalai - root - WARNING: Triton is not available, fallback to old kernel.                                                                                                                                                    
>>> 
```

meyersbs committed:
New package: py-fastfold

meyersbs committed:
Fix dependencies for py-fastfold

meyersbs commented:
@spackbot fix style

meyersbs committed:
Merge branch 'develop' into meyersbs/py-fastfold

meyersbs commented:
@spackbot fix style

adamjstewart commented on var/spack/repos/builtin/packages/py-fastfold/package.py:
```diff
@@ -0,0 +1,40 @@
+    # From import errors:
+    depends_on("py-setproctitle",               type="run")
+    depends_on("py-pdbfixer",                   type="run")
+    depends_on("py-pytorch-lightning",          type="run")
```
Most of these `type="run"` should be `type=("build", "run")` so that `spack install --test=root py-fastfold` passes.

meyersbs replied:
Can you define "most"? The rules on `type="run"` vs `type=("build", "run")` are still a bit fuzzy to me.

adamjstewart replied:
`spack install --test=root` attempts to import all installed files. So if importing a file imports one of these deps, it should be build/run. Honestly I would just make all of them build/run (except setuptools).

meyersbs replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/py-fastfold/package.py:
```diff
@@ -0,0 +1,40 @@
+    depends_on("py-einops",                     type=("build", "run"))
+    depends_on("py-colossalai",                 type=("build", "run"))
+    # From environment.yml (to appease import errors):
+    depends_on("py-scipy@1.7.1",                type="run")
+    depends_on("py-dm-tree@0.1.6",              type="run")
```
Would be great to relax more of these strict version requirements if possible

meyersbs replied:
I'll see what I can do.

meyersbs replied:
Versions relaxed; builds, loads, and passes import tests.

adamjstewart reviewed (commented):
@aweits may be interested in this, it's very similar to `py-alphafold`

meyersbs committed:
[py-fastfold] fix types

meyersbs committed:
[py-fastfold] relax strict versions

meyersbs committed:
Merge branch 'meyersbs/py-fastfold' of github.com:meyersbs/spack into meyersbs/py-fastfold

meyersbs committed:
Merge branch 'develop' into meyersbs/py-fastfold

meyersbs commented:
@spackbot fix style

meyersbs committed:
Fix spack style manually

meyersbs committed:
Merge branch 'develop' into meyersbs/py-fastfold

adamjstewart commented on var/spack/repos/builtin/packages/py-fastfold/package.py:
```diff
@@ -0,0 +1,40 @@
+    version("0.2.0", sha256="6760dbae9809b8b26219c9477489d34325807be504098901d0375fbdc3103f88")
+
+    # From README:
+    depends_on("python@3.8:3.9", type=("build", "run"))
```
Do Python 3.10 or 3.11 work? 3.10 is our default version

meyersbs replied:
Building now

adamjstewart commented on var/spack/repos/builtin/packages/py-fastfold/package.py:
```diff
@@ -0,0 +1,40 @@
+    depends_on("py-ray@2.0.0:", type=("build", "run"))  # Req relaxed (was @2.0.0)
+    depends_on("py-pandas", type=("build", "run"))
+    depends_on("openmm@7.7.0:+cuda", type=("build", "run"))
+    depends_on("py-requests@2.26.0", type=("build", "run"))
```
Does newer requests work?

meyersbs replied:
Missed that one; it should because it built/imported @2.28.0 when I typoed a while back. I'll double check.

meyersbs committed:
[py-fastfold] relax python and py-requests upper bounds

meyersbs commented:
@adamjstewart 
```
$ spack install py-fastfold ^python@3.10:
...
==> Installing py-fastfold-0.2.0-dzjpsh4nf5titspckvfrqum5ohi7tjzc
==> No binary for py-fastfold-0.2.0-dzjpsh4nf5titspckvfrqum5ohi7tjzc found: installing from source
==> Using cached archive: /home/bsmits/spack/var/spack/cache/_source-cache/archive/67/6760dbae9809b8b26219c9477489d34325807be504098901d0375fbdc3103f88.tar.gz
==> No patches needed for py-fastfold
==> py-fastfold: Executing phase: 'install'
==> py-fastfold: Successfully installed py-fastfold-0.2.0-dzjpsh4nf5titspckvfrqum5ohi7tjzc
  Fetch: 0.02s.  Build: 3m 13.16s.  Total: 3m 13.18s.
...
$ spack load py-fastfold
$ python3
Python 3.10.2 (main, Dec  8 2022, 8:55:28) [GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastfold; from fastfold.config import * ; from fastfold.workflow import * ; from fastfold.utils import * ; from fastfold.relax import * ; from fastfold.model import * ; from fastfold.distributed import * ; from fastfold.data import * ; from fastfold.common import * ; from fastfold.data.tools import * ; from fastfold.model.fastnn import * ; from fastfold.model.hub import * ; from fastfold.model.nn import * ; from fastfold.model.nn.embedders import * ; from fastfold.model.nn.msa import * ; from fastfold.model.nn.dropout import * ; from fastfold.model.hub.alphafold import * ; from fastfold.model.fastnn.embedders import * ; from fastfold.model.fastnn.evoformer import * ; from fastfold.utils.geometry import * ; from fastfold.utils.geometry.rigid_matrix_vector import * ; from fastfold.workflow.factory import * ; from fastfold.workflow.template import *; from fastfold.workflow.factory.hhsearch import *; from fastfold.data.data_modules import *; from fastfold.relax.amber_minimize import *
[12/08/22 08:49:57] WARNING  colossalai - root - WARNING: Triton is not available, fallback to old kernel.                                                                                    
                    WARNING  colossalai - root - WARNING: Triton is not available, fallback to old kernel.                                                                                    
[12/08/22 08:50:21] INFO     colossalai - numexpr.utils - INFO: Note: NumExpr detected 36 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.                              
>>> 
```

meyersbs committed:
Merge branch 'develop' into meyersbs/py-fastfold

adamjstewart commented on var/spack/repos/builtin/packages/py-fastfold/package.py:
```diff
@@ -0,0 +1,42 @@
+    depends_on("python@3.8:", type=("build", "run"))  # Req Relaxed (was @3.8:3.9)
+    depends_on("cuda@11.1:", type=("build", "run"))
+    depends_on("py-torch@1.10:+cuda", type=("build", "run"))
+    # Needed or py-torch fails to build when python@3.10:
+    depends_on("cudnn@8.0.5:", type=("build", "run"), when="python@3.10:")
```
This should go in the py-torch package then. Can you open a separate PR for that and remove it here? A link to the issue or discussion of why this is necessary in the PR would be helpful, it isn't immediately clear to me why this would matter.

meyersbs replied:
I'm not clear on why it matters, either. When I did `spack install py-fastfold ^python@3.10:` without that line, spack grabbed `cudnn@8.0.4.30-11.1` and the py-torch install exploded. I'll talk to @qwertos and see if we can debug further.

meyersbs replied:
My build environment was a bit confused, so I started fresh and py-torch installed just fine under `py-fastfold ^python@3.10:`, but the install fails for node-js (py-ray depends on npm):

```
==> Installing node-js-14.15.1-tjt2cpgf4udzxncylxajf7dbf4vhbuex
==> No binary for node-js-14.15.1-tjt2cpgf4udzxncylxajf7dbf4vhbuex found: installing from source
==> Using cached archive: /dev/shm/bsmits_spack/var/spack/cache/_source-cache/archive/a1/a1120472bf55aea745287693a6651e16973e1008c9d6107df350126adf9716fe.tar.gz
==> No patches needed for node-js
==> node-js: Executing phase: 'configure'
==> node-js: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    '/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/gmake-4.3-jjnsmedwhet6aiql2iultpawqztfao5z/bin/make' '-j16'

3 errors found in build log:
  >> 3      /scratch.local/spack/stage/bsmits/spack-stage-node-js-14.15.1-tjt2cpgf4udzxncylxajf7dbf4vhbuex/spack-src/configure.py:17: DeprecationWarning: The distutils package is deprecated
             and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
     4        from distutils.spawn import find_executable as which
     5      INFO: configure completed successfully
     6      ==> node-js: Executing phase: 'build'
     7      ==> [2022-12-12-09:46:48.840695] '/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/gmake-4.3-jjnsmedwhet6aiql2iultpawqztfao5z/bin/make' '-j16'
     8      /dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/gmake-4.3-jjnsmedwhet6aiql2iultpawqztfao5z/bin/make -C out BUILDTYPE=Release V=0
     9        touch /scratch.local/spack/stage/bsmits/spack-stage-node-js-14.15.1-tjt2cpgf4udzxncylxajf7dbf4vhbuex/spack-src/out/Release/obj.target/tools/icu/icu_implementation.stamp

     ...

     273        from jinja2.defaults import BLOCK_START_STRING, \
     274      File "/scratch.local/spack/stage/bsmits/spack-stage-node-js-14.15.1-tjt2cpgf4udzxncylxajf7dbf4vhbuex/spack-src/deps/v8/third_party/jinja2/defaults.py", line 32, in <module>
     275        from jinja2.tests import TESTS as DEFAULT_TESTS
     276      File "/scratch.local/spack/stage/bsmits/spack-stage-node-js-14.15.1-tjt2cpgf4udzxncylxajf7dbf4vhbuex/spack-src/deps/v8/third_party/jinja2/tests.py", line 13, in <module>
     277        from collections import Mapping
     278    ImportError: cannot import name 'Mapping' from 'collections' (/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/python-3.10.8-aophm7x3fgzm3xbtwcm4fuv4ztymujdu
            /lib/python3.10/collections/__init__.py)
  >> 279    make[1]: *** [tools/v8_gypfiles/v8_base_without_compiler.target.mk:30: 0309945459d2e6e8c485d40e87eb5f35969af51e.intermediate] Error 1
     280    make[1]: *** Waiting for unfinished jobs....
     281    rm 0309945459d2e6e8c485d40e87eb5f35969af51e.intermediate
  >> 282    make: *** [Makefile:104: node] Error 2
```

Python's `collections.Mapping` is [deprecated](https://docs.python.org/3.9/library/collections.html) in `python@3.10:`. Forcing a newer version of `node-js@15.3.0` results in the same error.

I'm not sure what the best way to handle this is.

adamjstewart replied:
I'm working on a fix for this, I'll ping you when it's ready.

adamjstewart replied:
Fixed in #34478

meyersbs committed:
[py-fastfold] remove bad/unecessary depends_on

meyersbs committed:
Merge branch 'meyersbs/py-fastfold' of github.com:meyersbs/spack into meyersbs/py-fastfold

adamjstewart reviewed (approved):
LGTM, would be good to get your confirmation that everything works now that node-js is fixed.

meyersbs commented:
> LGTM, would be good to get your confirmation that everything works now that node-js is fixed.

I'll give it a test tomorrow

meyersbs committed:
Merge branch 'develop' into meyersbs/py-fastfold

meyersbs commented:
@adamjstewart As soon as #34478 is merged, I'll test another build of py-fastfold.

meyersbs committed:
Merge branch 'develop' into meyersbs/py-fastfold

meyersbs commented:
@adamjstewart #34478 worked, but now I get a different error:

```
$ spack install py-fastfold ^python@3.10
...
==> Installing npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb
==> No binary for npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb found: installing from source
==> Using cached archive: /dev/shm/bsmits_spack/var/spack/cache/_source-cache/archive/1e/1e0e880ce0d5adf0120fb3f92fc8e5ea5bac73681d37282615d074ff670f7703.tgz
==> Using cached archive: /dev/shm/bsmits_spack/var/spack/cache/_source-cache/archive/bb/bbc0e137e17a63676efc97a0e3b1fcf101498a1c2c01c3341cd9491f248711b8.tgz
==> Using cached archive: /dev/shm/bsmits_spack/var/spack/cache/_source-cache/archive/16/168b394fbca60ea81dc84b1824466df96246b9eb4d671c2541f55f408a264b4c.tgz
==> Moving resource stage
	source: /scratch.local/spack/stage/bsmits/resource-node-gyp-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/
	destination: /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/node-gyp/package
==> Moving resource stage
	source: /scratch.local/spack/stage/bsmits/resource-env-paths-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/
	destination: /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/env-paths/package
==> Ran patch() for npm
==> npm: Executing phase: 'configure'
==> npm: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

1 error found in build log:
     222    npm ERR!
     223    npm ERR! Failed at the npm-cli-docs@0.1.0 build:static script.
     224    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
     225    
     226    npm ERR! A complete log of this run can be found in:
     227    npm ERR!     /home/bsmits/.npm/_logs/2022-12-19T13_31_09_574Z-debug.log
  >> 228    make: *** [htmldocs] Error 1

See build log for details:
  /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-build-out.txt
```

<details>
<summary>/home/bsmits/.npm/_logs/2022-12-19T13_31_09_574Z-debug.log</summary>

```
$ cat /home/bsmits/.npm/_logs/2022-12-19T13_31_09_574Z-debug.log
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/node-js-14.15.1-emqkf6mlix5wq4r4pgyieepjodqkn4ua/bin/node',
1 verbose cli   '/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/bin/npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'build:static',
1 verbose cli   'echo'
1 verbose cli ]
2 info using npm@6.14.9
3 info using node@v14.15.1
4 verbose run-script [ 'prebuild:static', 'build:static', 'postbuild:static' ]
5 info lifecycle npm-cli-docs@0.1.0~prebuild:static: npm-cli-docs@0.1.0
6 info lifecycle npm-cli-docs@0.1.0~build:static: npm-cli-docs@0.1.0
7 verbose lifecycle npm-cli-docs@0.1.0~build:static: unsafe-perm in lifecycle true
8 verbose lifecycle npm-cli-docs@0.1.0~build:static: PATH: /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/node_modules/npm-lifecycle/node-gyp-bin:/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/.bin:/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/node-js-14.15.1-emqkf6mlix5wq4r4pgyieepjodqkn4ua/bin:/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libvips-8.13.3-ackl6sxvv4tma4wqqkyhlhdpmxve6wgf/bin:/dev/shm/bsmits_spack/lib/spack/env/gcc:/dev/shm/bsmits_spack/lib/spack/env/case-insensitive:/dev/shm/bsmits_spack/lib/spack/env:/home/bsmits/.spack/bootstrap/store/linux-rhel7-x86_64/gcc-4.8.5/patchelf-0.13.1-vk2w2yu27hidlkntagtoqiz4jtqqsuuf/bin:/dev/shm/bsmits_spack/bin:/tools/bin/blindfold:/tools/spack/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/tools/bin:/home/bsmits/bin
9 verbose lifecycle npm-cli-docs@0.1.0~build:static: CWD: /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs
10 silly lifecycle npm-cli-docs@0.1.0~build:static: Args: [ '-c', 'GATSBY_IS_STATIC=true gatsby build --prefix-paths "echo"' ]
11 silly lifecycle npm-cli-docs@0.1.0~build:static: Returned: code: 1  signal: null
12 info lifecycle npm-cli-docs@0.1.0~build:static: Failed to exec build:static script
13 verbose stack Error: npm-cli-docs@0.1.0 build:static: `GATSBY_IS_STATIC=true gatsby build --prefix-paths "echo"`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:315:20)
13 verbose stack     at ChildProcess.<anonymous> (/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:315:20)
13 verbose stack     at maybeClose (internal/child_process.js:1048:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)
14 verbose pkgid npm-cli-docs@0.1.0
15 verbose cwd /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs
16 verbose Linux 5.10.155-1.el7.x86_64
17 verbose argv "/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/node-js-14.15.1-emqkf6mlix5wq4r4pgyieepjodqkn4ua/bin/node" "/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/bin/npm-cli.js" "run" "build:static" "echo"
18 verbose node v14.15.1
19 verbose npm  v6.14.9
20 error code ELIFECYCLE
21 error errno 1
22 error npm-cli-docs@0.1.0 build:static: `GATSBY_IS_STATIC=true gatsby build --prefix-paths "echo"`
22 error Exit status 1
23 error Failed at the npm-cli-docs@0.1.0 build:static script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

</details>

<details>
<summary>/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-build-out.txt</summary>

```
==> npm: Executing phase: 'configure'
==> [2022-12-19-08:30:21.075173] './configure' '--prefix=/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb'
==> npm: Executing phase: 'build'
==> [2022-12-19-08:30:21.164037] 'make' '-j16'
node bin/npm-cli.js install marked --no-global --no-timing --no-save
node bin/npm-cli.js install marked-man --no-global --no-timing --no-save
cd docs && node ../bin/npm-cli.js install && \
node ../bin/npm-cli.js run build:static echo>&2 && \
rm -rf node_modules .cache public/*js public/*json public/404* public/page-data public/manifest*
+ marked@0.7.0
added 2 packages from 2 contributors, updated 1 package and audited 501 packages in 2.1s
npm WARN marked-man@0.7.0 requires a peer of marked@^0.7.0 but none is installed. You must install peer dependencies yourself.


3 packages are looking for funding
  run `npm fund` for details

found 87 vulnerabilities (1 low, 10 moderate, 62 high, 14 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
+ marked-man@0.7.0
added 2 packages from 2 contributors, updated 1 package and audited 501 packages in 2.23s

3 packages are looking for funding
  run `npm fund` for details

found 85 vulnerabilities (1 low, 10 moderate, 60 high, 14 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
node scripts/docs-build.js docs/content/cli-commands/npm-access.md man/man1/npm-access.1
node scripts/docs-build.js docs/content/cli-commands/npm-adduser.md man/man1/npm-adduser.1
node scripts/docs-build.js docs/content/cli-commands/npm-audit.md man/man1/npm-audit.1
node scripts/docs-build.js docs/content/cli-commands/npm-bin.md man/man1/npm-bin.1
node scripts/docs-build.js docs/content/cli-commands/npm-bugs.md man/man1/npm-bugs.1
node scripts/docs-build.js docs/content/cli-commands/npm-build.md man/man1/npm-build.1
node scripts/docs-build.js docs/content/cli-commands/npm-bundle.md man/man1/npm-bundle.1
node scripts/docs-build.js docs/content/cli-commands/npm-cache.md man/man1/npm-cache.1
node scripts/docs-build.js docs/content/cli-commands/npm-ci.md man/man1/npm-ci.1
node scripts/docs-build.js docs/content/cli-commands/npm-completion.md man/man1/npm-completion.1
node scripts/docs-build.js docs/content/cli-commands/npm-config.md man/man1/npm-config.1
node scripts/docs-build.js docs/content/cli-commands/npm-dedupe.md man/man1/npm-dedupe.1
node scripts/docs-build.js docs/content/cli-commands/npm-deprecate.md man/man1/npm-deprecate.1
node scripts/docs-build.js docs/content/cli-commands/npm-dist-tag.md man/man1/npm-dist-tag.1
node scripts/docs-build.js docs/content/cli-commands/npm-docs.md man/man1/npm-docs.1
node scripts/docs-build.js docs/content/cli-commands/npm-doctor.md man/man1/npm-doctor.1
node scripts/docs-build.js docs/content/cli-commands/npm-edit.md man/man1/npm-edit.1
node scripts/docs-build.js docs/content/cli-commands/npm-explore.md man/man1/npm-explore.1
node scripts/docs-build.js docs/content/cli-commands/npm-fund.md man/man1/npm-fund.1
node scripts/docs-build.js docs/content/cli-commands/npm-help-search.md man/man1/npm-help-search.1
node scripts/docs-build.js docs/content/cli-commands/npm-help.md man/man1/npm-help.1
node scripts/docs-build.js docs/content/cli-commands/npm-hook.md man/man1/npm-hook.1
node scripts/docs-build.js docs/content/cli-commands/npm-init.md man/man1/npm-init.1
node scripts/docs-build.js docs/content/cli-commands/npm-install-ci-test.md man/man1/npm-install-ci-test.1
node scripts/docs-build.js docs/content/cli-commands/npm-install-test.md man/man1/npm-install-test.1
node scripts/docs-build.js docs/content/cli-commands/npm-install.md man/man1/npm-install.1
node scripts/docs-build.js docs/content/cli-commands/npm-link.md man/man1/npm-link.1
node scripts/docs-build.js docs/content/cli-commands/npm-logout.md man/man1/npm-logout.1
node scripts/docs-build.js docs/content/cli-commands/npm-ls.md man/man1/npm-ls.1
node scripts/docs-build.js docs/content/cli-commands/npm-org.md man/man1/npm-org.1
node scripts/docs-build.js docs/content/cli-commands/npm-outdated.md man/man1/npm-outdated.1
node scripts/docs-build.js docs/content/cli-commands/npm-owner.md man/man1/npm-owner.1
node scripts/docs-build.js docs/content/cli-commands/npm-pack.md man/man1/npm-pack.1
node scripts/docs-build.js docs/content/cli-commands/npm-ping.md man/man1/npm-ping.1
node scripts/docs-build.js docs/content/cli-commands/npm-prefix.md man/man1/npm-prefix.1
node scripts/docs-build.js docs/content/cli-commands/npm-profile.md man/man1/npm-profile.1
node scripts/docs-build.js docs/content/cli-commands/npm-prune.md man/man1/npm-prune.1
node scripts/docs-build.js docs/content/cli-commands/npm-publish.md man/man1/npm-publish.1
node scripts/docs-build.js docs/content/cli-commands/npm-rebuild.md man/man1/npm-rebuild.1
node scripts/docs-build.js docs/content/cli-commands/npm-repo.md man/man1/npm-repo.1
node scripts/docs-build.js docs/content/cli-commands/npm-restart.md man/man1/npm-restart.1
node scripts/docs-build.js docs/content/cli-commands/npm-root.md man/man1/npm-root.1
node scripts/docs-build.js docs/content/cli-commands/npm-run-script.md man/man1/npm-run-script.1
node scripts/docs-build.js docs/content/cli-commands/npm-search.md man/man1/npm-search.1
node scripts/docs-build.js docs/content/cli-commands/npm-shrinkwrap.md man/man1/npm-shrinkwrap.1
node scripts/docs-build.js docs/content/cli-commands/npm-star.md man/man1/npm-star.1
node scripts/docs-build.js docs/content/cli-commands/npm-stars.md man/man1/npm-stars.1
node scripts/docs-build.js docs/content/cli-commands/npm-start.md man/man1/npm-start.1
node scripts/docs-build.js docs/content/cli-commands/npm-stop.md man/man1/npm-stop.1
node scripts/docs-build.js docs/content/cli-commands/npm-team.md man/man1/npm-team.1
node scripts/docs-build.js docs/content/cli-commands/npm-test.md man/man1/npm-test.1
node scripts/docs-build.js docs/content/cli-commands/npm-token.md man/man1/npm-token.1
node scripts/docs-build.js docs/content/cli-commands/npm-uninstall.md man/man1/npm-uninstall.1
node scripts/docs-build.js docs/content/cli-commands/npm-unpublish.md man/man1/npm-unpublish.1
node scripts/docs-build.js docs/content/cli-commands/npm-update.md man/man1/npm-update.1
node scripts/docs-build.js docs/content/cli-commands/npm-version.md man/man1/npm-version.1
node scripts/docs-build.js docs/content/cli-commands/npm-view.md man/man1/npm-view.1
node scripts/docs-build.js docs/content/cli-commands/npm-whoami.md man/man1/npm-whoami.1
node scripts/docs-build.js docs/content/cli-commands/npm.md man/man1/npm.1
node scripts/docs-build.js README.md man/man1/npm-README.1
node scripts/docs-build.js docs/content/configuring-npm/folders.md man/man5/folders.5
node scripts/docs-build.js docs/content/configuring-npm/install.md man/man5/install.5
node scripts/docs-build.js docs/content/configuring-npm/npmrc.md man/man5/npmrc.5
node scripts/docs-build.js docs/content/configuring-npm/package-json.md man/man5/package-json.5
node scripts/docs-build.js docs/content/configuring-npm/package-lock-json.md man/man5/package-lock-json.5
node scripts/docs-build.js docs/content/configuring-npm/package-locks.md man/man5/package-locks.5
node scripts/docs-build.js docs/content/configuring-npm/shrinkwrap-json.md man/man5/shrinkwrap-json.5
node scripts/docs-build.js docs/content/using-npm/config.md man/man7/config.7
node scripts/docs-build.js docs/content/using-npm/developers.md man/man7/developers.7
node scripts/docs-build.js docs/content/using-npm/disputes.md man/man7/disputes.7
node scripts/docs-build.js docs/content/using-npm/orgs.md man/man7/orgs.7
node scripts/docs-build.js docs/content/using-npm/registry.md man/man7/registry.7
node scripts/docs-build.js docs/content/using-npm/removal.md man/man7/removal.7
node scripts/docs-build.js docs/content/using-npm/scope.md man/man7/scope.7
node scripts/docs-build.js docs/content/using-npm/scripts.md man/man7/scripts.7
node scripts/docs-build.js docs/content/using-npm/semver.md man/man7/semver.7

> sharp@0.23.4 install /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/sharp
> (node install/libvips && node install/dll-copy && prebuild-install) || (node-gyp rebuild && node install/dll-copy)

info sharp Detected globally-installed libvips v8.13.3
info sharp Building from source via node-gyp
make[1]: Entering directory `/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/sharp/build'
make[1]: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
  TOUCH Release/obj.target/libvips-cpp.stamp
  CXX(target) Release/obj.target/sharp/src/common.o
  CXX(target) Release/obj.target/sharp/src/metadata.o
  CXX(target) Release/obj.target/sharp/src/stats.o
  CXX(target) Release/obj.target/sharp/src/operations.o
  CXX(target) Release/obj.target/sharp/src/pipeline.o
  CXX(target) Release/obj.target/sharp/src/sharp.o
  CXX(target) Release/obj.target/sharp/src/utilities.o
  SOLINK_MODULE(target) Release/obj.target/sharp.node
  COPY Release/sharp.node
make[1]: Leaving directory `/scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/sharp/build'

> core-js@2.6.11 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon: 
> https://opencollective.com/core-js 
> https://www.patreon.com/zloirock 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> core-js-pure@3.6.2 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/core-js-pure
> node -e "try{require('./postinstall')}catch(e){}"


> gatsby-telemetry@1.1.46 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/gatsby-telemetry
> node src/postinstall.js || true


> cwebp-bin@5.1.0 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/cwebp-bin
> node lib/install.js

  ✔ cwebp pre-build test passed successfully

> mozjpeg@6.0.1 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/mozjpeg
> node lib/install.js

  ✔ mozjpeg pre-build test passed successfully

> pngquant-bin@5.0.2 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/pngquant-bin
> node lib/install.js

  ✔ pngquant pre-build test passed successfully

> gatsby-cli@2.8.23 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/gatsby/node_modules/gatsby-cli
> node scripts/postinstall.js


> gatsby@2.18.18 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/gatsby
> node scripts/postinstall.js


> styled-components@4.4.1 postinstall /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs/node_modules/styled-components
> node ./scripts/postinstall.js || exit 0

Use styled-components at work? Consider supporting our development efforts at https://opencollective.com/styled-components
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.11 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.11: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.2 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 2176 packages from 1039 contributors and audited 2249 packages in 42.195s

54 packages are looking for funding
  run `npm fund` for details

found 346 vulnerabilities (19 low, 120 moderate, 174 high, 33 critical)
  run `npm audit fix` to fix them, or `npm audit` for details

> npm-cli-docs@0.1.0 build:static /scratch.local/spack/stage/bsmits/spack-stage-npm-6.14.9-hmdybw7yf5uvthrdyavlrduymz4ktxmb/spack-src/docs
> GATSBY_IS_STATIC=true gatsby build --prefix-paths "echo"

success open and validate gatsby-configs - 0.045s
success load plugins - 0.454s
success onPreInit - 0.005s
success delete html and css files from previous builds - 0.024s
success initialize cache - 0.007s
success copy gatsby files - 0.020s
success onPreBootstrap - 0.004s
success createSchemaCustomization - 0.012s
success source and transform nodes - 0.288s
success building schema - 0.280s
warn code block or inline code language not specified in markdown. applying generic code block
success createPages - 0.793s
success createPagesStatefully - 0.206s
success onPreExtractQueries - 0.001s
success update schema - 0.031s
success extract queries from components - 0.116s
success write out requires - 0.004s
success write out redirect data - 0.001s

 ERROR #11321  PLUGIN

"gatsby-plugin-manifest" threw an error while running the onPostBootstrap lifecycle:

Input file contains unsupported image format



  Error: Input file contains unsupported image format

not finished onPostBootstrap - 0.056s
not finished Build manifest and related icons - 0.069s
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! npm-cli-docs@0.1.0 build:static: `GATSBY_IS_STATIC=true gatsby build --prefix-paths "echo"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the npm-cli-docs@0.1.0 build:static script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/bsmits/.npm/_logs/2022-12-19T13_31_09_574Z-debug.log
make: *** [htmldocs] Error 1
```
</details>

Doing `spack install npm@6.14.9` on it's own gives the same error.

adamjstewart commented:
@cosmicexplorer do you know anything? The latest version of npm in Spack is quite old, and may not be compatible with newer node-js.

meyersbs committed:
Merge branch 'develop' into meyersbs/py-fastfold

adamjstewart commented:
Any objections to merging this PR as is? We can work on npm in a separate PR.

meyersbs commented:
That's your call. I just tried building in a fresh spack environment again and got a different npm error:

```
==> Installing npm-6.14.9-gu6b5bddoyu2e23lgjehcu3clzerx6ym
==> npm exists in binary cache but with different hash
==> No binary for npm-6.14.9-gu6b5bddoyu2e23lgjehcu3clzerx6ym found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/1e/1e0e880ce0d5adf0120fb3f92fc8e5ea5bac73681d37282615d074ff670f7703.tgz
==> Fetching https://mirror.spack.io/_source-cache/archive/bb/bbc0e137e17a63676efc97a0e3b1fcf101498a1c2c01c3341cd9491f248711b8.tgz
==> Fetching https://mirror.spack.io/_source-cache/archive/16/168b394fbca60ea81dc84b1824466df96246b9eb4d671c2541f55f408a264b4c.tgz
==> Moving resource stage
        source: /dev/shm/spack/stage/bsmits/resource-node-gyp-gu6b5bddoyu2e23lgjehcu3clzerx6ym/spack-src/
        destination: /dev/shm/spack/stage/bsmits/spack-stage-npm-6.14.9-gu6b5bddoyu2e23lgjehcu3clzerx6ym/spack-src/node-gyp/package
==> Moving resource stage
        source: /dev/shm/spack/stage/bsmits/resource-env-paths-gu6b5bddoyu2e23lgjehcu3clzerx6ym/spack-src/
        destination: /dev/shm/spack/stage/bsmits/spack-stage-npm-6.14.9-gu6b5bddoyu2e23lgjehcu3clzerx6ym/spack-src/env-paths/package
==> Ran patch() for npm
==> npm: Executing phase: 'configure'
==> npm: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

7 errors found in build log:
     116    In file included from /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-local-handle.h:12,
     117                     from /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-array-buffer.h:12,
     118                     from /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8.h:24,
     119                     from /home/bsmits/.cache/node-gyp/18.12.1/include/node/node.h:73,
     120                     from ../src/common.cc:22:
     121    /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-internal.h: In function 'void v8::internal::PerformCastCheck(T*)':
  >> 122    /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-internal.h:646:38: error: 'remove_cv_t' is not a member of 'std'; did you mean 'remove_cv'?
     123      646 |             !std::is_same<Data, std::remove_cv_t<T>>::value>::Perform(data);
     124          |                                      ^~~~~~~~~~~
     125          |                                      remove_cv
  >> 126    /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-internal.h:646:38: error: 'remove_cv_t' is not a member of 'std'; did you mean 'remove_cv'?
     127      646 |             !std::is_same<Data, std::remove_cv_t<T>>::value>::Perform(data);
     128          |                                      ^~~~~~~~~~~
     129          |                                      remove_cv
  >> 130    /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-internal.h:646:50: error: template argument 2 is invalid
     131      646 |             !std::is_same<Data, std::remove_cv_t<T>>::value>::Perform(data);
     132          |                                                  ^
  >> 133    /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-internal.h:646:63: error: '::Perform' has not been declared
     134      646 |             !std::is_same<Data, std::remove_cv_t<T>>::value>::Perform(data);
     135          |                                                               ^~~~~~~
     136    In file included from ../src/common.cc:24:
     137    ../../nan/nan.h: In function 'void Nan::SetAccessor(v8::Local<v8::ObjectTemplate>, v8::Local<v8::String>, Nan::GetterCallback, Nan::SetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, Nan::imp::Sig)':
     138    ../../nan/nan.h:2543:16: warning: 'void v8::ObjectTemplate::SetAccessor(v8::Local<v8::Name>, v8::AccessorNameGetterCallback, v8::AccessorNameSetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, v8::Local<v8::Access
            orSignature>, v8::SideEffectType, v8::SideEffectType)' is deprecated: Do signature check in accessor [-Wdeprecated-declarations]
     139     2543 |     , signature);

     ...

     145    /home/bsmits/.cache/node-gyp/18.12.1/include/node/v8-template.h:838:8: note: declared here
     146      838 |   void SetAccessor(
     147          |        ^~~~~~~~~~~
     148    In file included from ../../nan/nan.h:2884,
     149                     from ../src/common.cc:24:
     150    ../../nan/nan_typedarray_contents.h: In constructor 'Nan::TypedArrayContents<T>::TypedArrayContents(v8::Local<v8::Value>)':
  >> 151    ../../nan/nan_typedarray_contents.h:34:43: error: 'class v8::ArrayBuffer' has no member named 'GetContents'
     152       34 |       data   = static_cast<char*>(buffer->GetContents().Data()) + byte_offset;
     153          |                                           ^~~~~~~~~~~
  >> 154    make[1]: *** [Release/obj.target/sharp/src/common.o] Error 1
     155    make[1]: Leaving directory `/dev/shm/spack/stage/bsmits/spack-stage-npm-6.14.9-gu6b5bddoyu2e23lgjehcu3clzerx6ym/spack-src/docs/node_modules/sharp/build'
     156    gyp ERR! build error
     157    gyp ERR! stack Error: `make` failed with exit code: 2
     158    gyp ERR! stack     at ChildProcess.onExit (/dev/shm/spack/stage/bsmits/spack-stage-npm-6.14.9-gu6b5bddoyu2e23lgjehcu3clzerx6ym/spack-src/node_modules/node-gyp/lib/build.js:194:23)
     159    gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
     160    gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:291:12)

     ...

     176    npm ERR!
     177    npm ERR! Failed at the sharp@0.23.4 install script.
     178    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
     179    
     180    npm ERR! A complete log of this run can be found in:
     181    npm ERR!     /home/bsmits/.npm/_logs/2023-01-12T15_31_55_474Z-debug.log
  >> 182    make: *** [htmldocs] Error 1
```

adamjstewart commented:
I'm pretty sure the recipe for fastfold is right, but we can always open up a follow-up PR if we need to correct anything. Unfortunately I can't help much with npm, and there's no official maintainer for that package recipe. You could open an issue with the npm devs, but they will likely ask why you're trying to build such an old version of npm. A better solution would be to ask them how to build the latest npm from source so we can do that instead. I have a feeling they'll tell us we should combine node-js and npm into a single package like homebrew does. Maybe @cosmicexplorer has thoughts on this.
